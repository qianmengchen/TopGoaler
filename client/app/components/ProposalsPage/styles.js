import { StyleSheet } from 'react-native';

const list = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'stretch'
  },
  title: {
    textAlign: 'center',
    fontSize: 20,
    fontWeight: 'bold',
    marginTop: 10
  }
});

const frequency = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 10
  },
  textgroup: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  }
});

const vote = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
    marginTop: 10,
    marginHorizontal: 80
  },
  unvoted: {
    backgroundColor: 'red'
  },
  voted: {
    backgroundColor: 'grey'
  },
  buttonText: {
    color: '#fff',
    textAlign: 'center',
    fontSize: 17,
    paddingVertical: 10
  }
});

export { list, frequency, vote };
