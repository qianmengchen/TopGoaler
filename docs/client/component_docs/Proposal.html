<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: Proposal</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: Proposal</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>Proposal<span class="signature">()</span><span class="type-signature"></span></h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    

    
    <h4 class="name" id="Proposal"><span class="type-signature"></span>new Proposal<span class="signature">()</span><span class="type-signature"></span></h4>
    

    















    <h5 class="subsection-title">Properties:</h5>

    

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>key</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A unique number identifying this proposal.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>title</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The title of the new task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>subtitle</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">A short description of the new task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>period</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A type encoding indicating the day, week, or month period of a task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>pattern</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A value indicating the numbere of times a user can complete a task within a given period.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>handleVote</code></td>
            

            <td class="type">
            
                
<span class="param-type">function</span>


            
            </td>

            

            

            <td class="description last">A function allowing the user to vote on a specific this task and update the vote count.</td>
        </tr>

    
    </tbody>
</table>




<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="Proposal.js.html">Proposal.js</a>, <a href="Proposal.js.html#line19">line 19</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    

    

    

    

    

    

    

    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="Proposal.html">Proposal</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Sun Dec 02 2018 21:04:41 GMT-0800 (Pacific Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>