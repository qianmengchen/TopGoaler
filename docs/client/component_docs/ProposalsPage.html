<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: ProposalsPage</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: ProposalsPage</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>ProposalsPage<span class="signature">()</span><span class="type-signature"></span></h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    

    
    <h4 class="name" id="ProposalsPage"><span class="type-signature"></span>new ProposalsPage<span class="signature">()</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    This is the component that will show a list of proposal component currently in that channel.
</div>













    <h5 class="subsection-title">Properties:</h5>

    

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>channel_id</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A unique value identifying the channel.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>user_id</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A unique value identifying the user.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>proposals</code></td>
            

            <td class="type">
            
                
<span class="param-type">Component</span>


            
            </td>

            

            

            <td class="description last">The Proposals component  embedded within this ProposalsPage component.</td>
        </tr>

    
    </tbody>
</table>




<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="ProposalsPage_ProposalsPage.js.html">ProposalsPage/ProposalsPage.js</a>, <a href="ProposalsPage_ProposalsPage.js.html#line6">line 6</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    

    

    

    

    

    

    

    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="App.html">App</a></li><li><a href="AuthNav.html">AuthNav</a></li><li><a href="ChannelIcon.html">ChannelIcon</a></li><li><a href="ChannelIconList.html">ChannelIconList</a></li><li><a href="ChannelListPage.html">ChannelListPage</a></li><li><a href="ChannelMemberView.html">ChannelMemberView</a></li><li><a href="ChannelOverview.html">ChannelOverview</a></li><li><a href="ChannelPublicView.html">ChannelPublicView</a></li><li><a href="Feed.html">Feed</a></li><li><a href="MemberList.html">MemberList</a></li><li><a href="NewChannelPage.html">NewChannelPage</a></li><li><a href="NewTaskPage.html">NewTaskPage</a></li><li><a href="PerformancePage.html">PerformancePage</a></li><li><a href="Profile.html">Profile</a></li><li><a href="Proposal.html">Proposal</a></li><li><a href="ProposalsPage.html">ProposalsPage</a></li><li><a href="SignIn.html">SignIn</a></li><li><a href="SignUp.html">SignUp</a></li><li><a href="TaskChecklist.html">TaskChecklist</a></li><li><a href="TaskDetail.html">TaskDetail</a></li><li><a href="TaskEntry.html">TaskEntry</a></li><li><a href="TaskListPage.html">TaskListPage</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Mon Dec 03 2018 02:45:35 GMT-0800 (Pacific Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>