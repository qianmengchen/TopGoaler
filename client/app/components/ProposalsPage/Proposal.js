import React, { Component } from 'react';
import { Text, View, Slider, TouchableHighlight } from 'react-native';
import { Card } from 'react-native-elements';
import { frequency, vote } from './styles';

class Proposal extends Component {
  _handleVote() {
    this.props.handleVote(this.state.points);
  }

  constructor(props) {
    super(props);
    this.state = {
      points: 50
    };
  }

  render() {
    let { title, subtitle, period, pattern } = this.props;

    return (
      <Card title={title}>
        <Text>{subtitle}</Text>

        <View style={frequency.container}>
          <Text>
            Frequency: {pattern} {period}{' '}
          </Text>
        </View>

        <Slider
          step={50}
          minimumValue={50}
          maximumValue={350}
          value={this.state.points}
          onValueChange={value => this.setState({ points: value })}
        />

        {/* <Text>{this.state.points}</Text> */}

        <TouchableHighlight
          style={vote.container}
          underlayColor="#aaa"
          onPress={this._handleVote.bind(this)}
        >
          <View>
            <Text style={vote.buttonText}>
              {' '}
              Vote for {this.state.points} points!{' '}
            </Text>
          </View>
        </TouchableHighlight>
      </Card>
    );
  }
}

export default Proposal;
