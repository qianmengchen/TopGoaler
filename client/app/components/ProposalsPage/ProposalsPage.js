import React, { Component } from 'react';
import { ScrollView } from 'react-native';
import { list } from './styles';
import Proposal from './Proposal';

/**
 * This is the component that will show a list of proposal component currently in that channel.
 *
 * @class ProposalsPage
 * @property {number} channel_id - A unique value identifying the channel.
 * @property {number} user_id - A unique value identifying the user.
 * @property {Component} proposals - The Proposals component  embedded within this ProposalsPage component.
 */

class ProposalsPage extends Component {
  _handleVoteFactory = proposal_id => points => {
    this.props.handleVote(
      this.props.user_id,
      proposal_id,
      points,
      this.props.channel_id
    );
  };
  static navigationOptions = {
    title: 'Task Proposals'
  };
  render() {
    return (
      <ScrollView contentContainerStyle={list.container}>
        {this.props.proposals.map(id => {
          return (
            <Proposal
              key={id.id}
              title={id.title}
              subtitle={id.subtitle}
              period={id.period}
              pattern={id.pattern}
              voted={id.voted}
              points={id.points}
              handleVote={this._handleVoteFactory(id.id)}
            />
          );
        })}
      </ScrollView>
    );
  }
}

export default ProposalsPage;
