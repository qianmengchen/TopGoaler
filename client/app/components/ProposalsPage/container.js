import { connect } from 'react-redux';
import ProposalsPage from './ProposalsPage';
import { handleVote } from '../../actions';

const mapStateToProps = (state, ownProps) => {
  let channel_id = ownProps.navigation.state.params.channel_id;
  let user_id = state.auth.id;
  let voted_id = state.votes.filter(x => user_id == x.user_id);
  let voted_map = new Map(voted_id.map(x => [x.proposal_id, x.score]));
  return {
    channel_id,
    user_id: state.auth.id,
    proposals: state.proposals
      .filter(x => channel_id == +x.channel_id)
      .map(x => ({
        ...x,
        voted: voted_map.has(x.id),
        points: voted_map.get(x.id) || 0
      }))
  };
};

const mapDispatchToProps = dispatch => ({
  handleVote: (user_id, proposal_id, score) =>
    dispatch(handleVote(user_id, proposal_id, score))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProposalsPage);
