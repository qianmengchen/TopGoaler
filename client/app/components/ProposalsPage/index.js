import ProposalsPage from './container';

export { ProposalsPage };
