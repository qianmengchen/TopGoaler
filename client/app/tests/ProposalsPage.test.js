import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunkMiddleware from 'redux-thunk';
import sinon from 'sinon';
require('isomorphic-fetch');

import ProposalsPage from '../components/ProposalsPage/ProposalsPage';
import Proposal from '../components/ProposalsPage/Proposal';

const middlewares = [thunkMiddleware]; // you can mock any middlewares here if necessary
const mockStore = configureStore(middlewares);

const initialTaskState = {
  tasks: ['do frontend', 'do backend', 'do MVP ftw'],
  isFetching: false,
  currentTaskPage: 'ChannelListPage'
};

const proposals = [
  {
    id: 1,
    title: 'do 10 hard problems',
    channel_id: '2',
    subtitle: 'this is a proposal',
    period: 0,
    pattern: 1
  },
  {
    id: 2,
    title: 'go to gym',
    channel_id: '3',
    subtitle: 'this is a proposal',
    period: 1,
    pattern: 4
  }
];

describe('Testing Proposals Page', () => {
  const wrapper = shallow(<ProposalsPage proposals={proposals} />, {
    context: { store: mockStore(initialTaskState) }
  });
  const render = wrapper.dive();

  it('renders as expected', () => {
    expect(render).toMatchSnapshot();
  });
});

describe('Testing Proposal Component', () => {
  const mockHandleVote = jest.fn();
  // const mockHandleVote = ProposalsPage.prototype._handleVoteFactory;
  const handleVote = sinon.spy(Proposal.prototype, '_handleVote');

  const wrapper = shallow(<Proposal handleVote={mockHandleVote} />, {
    context: { store: mockStore(initialTaskState) }
  });
  const render = wrapper.dive();

  it('renders as expected', () => {
    expect(render).toMatchSnapshot();
  });

  it('correctly updates point states on value change', () => {
    const slider = wrapper.find('Slider');
    expect(slider.props().value).toBe(50);

    slider.simulate('valueChange', 100);
    expect(wrapper.state('points')).toBe(100);

    // re-render for updated state
    const slider2 = wrapper.find('Slider');
    expect(slider2.props().value).toBe(100);
  });

  it('invokes correct methods when pressing vote button', () => {
    wrapper.find('TouchableHighlight').simulate('press');

    expect(handleVote.calledOnce).toBe(true);
  });
});
