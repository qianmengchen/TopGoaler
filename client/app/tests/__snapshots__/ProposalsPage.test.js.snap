// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Proposal Component renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Themed.Card>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
    <Component
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Frequency: 
        unlimitedtime(s) per undefined
      </Text>
    </Component>
    <Slider
      disabled={false}
      maximumValue={350}
      minimumValue={50}
      onValueChange={[Function]}
      step={50}
      value={50}
    />
    <TouchableHighlight
      activeOpacity={0.85}
      delayPressOut={100}
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 5,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginHorizontal": 80,
            "marginTop": 10,
          },
          Object {
            "backgroundColor": "red",
          },
        ]
      }
      underlayColor="#aaa"
    >
      <Component>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#fff",
              "fontSize": 17,
              "paddingVertical": 10,
              "textAlign": "center",
            }
          }
        >
          Vote for 
          50
           points!
        </Text>
      </Component>
    </TouchableHighlight>
  </Themed.Card>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "Consumer": [Circular],
      "Provider": Object {
        "$$typeof": Symbol(react.provider),
        "_context": [Circular],
      },
      "_calculateChangedBits": null,
      "_changedBits": 0,
      "_currentRenderer": null,
      "_currentValue": undefined,
      "_defaultValue": undefined,
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_currentRenderer": null,
        "_currentValue": undefined,
        "_defaultValue": undefined,
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;

exports[`Testing Proposals Page renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ScrollViewMock
    contentContainerStyle={
      Object {
        "alignItems": "stretch",
        "justifyContent": "center",
      }
    }
  >
    <Proposal
      handleVote={[Function]}
      pattern={1}
      period={0}
      subtitle="this is a proposal"
      title="do 10 hard problems"
    />
    <Proposal
      handleVote={[Function]}
      pattern={4}
      period={1}
      subtitle="this is a proposal"
      title="go to gym"
    />
  </ScrollViewMock>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [
    undefined,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
